linalg_ops/convolution/generate_e2e_conv2d_tests.py

```diff
@@ -0,0 +1,739 @@
+        signature=signature,
+        import_declaration=import_declaration,
+        definition=func_definition,
+    )
+
+
+# Represents a call to a generated test function.
+@dataclasses.dataclass
+class TestCall:
+    function: MLIRFunction
+    op: str
+
+
+# Enumerates ways to initialize tensor buffer contents.
+@enum.unique
+class TensorGenerator(enum.Enum):
+    ZERO = "zero"  # Fill with zeros
+    RANDOM = "random"  # Fill with (deterministic) pseudorandom values.
+
+
+# Intentionally fixed seed! We want full reproducibility here, both across runs
+# and across machines.
+# Intentionally not shared with local_pseudorandom_state to limit the ways
+# in which shuffling testcases changes which random values are generated.
+pseudorandom_generator_seed = 1
+
+
+def contents_generator_tag(generator: TensorGenerator):
+    if generator == TensorGenerator.ZERO:
+        return ""
+    elif generator == TensorGenerator.RANDOM:
+        global pseudorandom_generator_seed
+        pseudorandom_generator_seed = pseudorandom_generator_seed + 1
+        return f"!tag:iree:fully_specified_pseudorandom {pseudorandom_generator_seed}"
+    else:
+        raise ValueError(generator)
+
+
+# Generate a 4d tensor function argument of the given size as `%name`.
+def generate_random_4d_tensor(
+    name: str,
+    tensor_shape: list,
+    element_type: typing.Union[InputElemTypeId, KernelElemTypeId],
+):
+    global pseudorandom_generator_seed
+    pseudorandom_generator_seed = pseudorandom_generator_seed + 1
+    return (
+        f"  %{name}_dim0 = arith.constant {tensor_shape[0]} : i64\n"
+        f"  %{name}_dim1 = arith.constant {tensor_shape[1]} : i64\n"
+        f"  %{name}_dim2 = arith.constant {tensor_shape[2]} : i64\n"
+        f"  %{name}_dim3 = arith.constant {tensor_shape[3]} : i64\n"
+        f"  %{name}_element_type = hal.element_type<{element_type.value}> : i32\n"
+        f"  %{name}_seed = arith.constant {pseudorandom_generator_seed} : i32\n"
+        f"  %{name} = call @conv2d_test.generate_random_tensor(%device, %{name}_dim0, %{name}_dim1, %{name}_dim2, %{name}_dim3, %{name}_element_type, %{name}_seed) : (!hal.device, i64, i64, i64, i64, i32, i32) -> !hal.buffer_view\n"
+    )
+
+
+call_id = 0
+
+
+def generate_call(
+    function: MLIRFunction,
+    input_type: InputElemTypeId,
+    input_layout: InputLayout,
+    kernel_type: KernelElemTypeId,
+    kernel_layout: KernelLayout,
+    conv2d_attr: ConvAttrs,
+    acc_type: AccElemTypeId,
+    shape: TestShape,
+):
+    global call_id
+    func_name = f"{function.name}_{shape.n}_{shape.c}_{shape.h}_{shape.w}_{shape.f}_{shape.kh}_{shape.kw}"
+    if shape.accumulate:
+        func_name = f"{func_name}_acc"
+    func_name = f"{func_name}_{call_id}"
+    call_id = call_id + 1
+
+        # layout of output tensor for checking correctness
+    layout = -1
+    description = f"Conv2d shape (NxCxHxWxFxKHxKW): {shape.n}x{shape.c}x{shape.h}x{shape.w}x{shape.f}x{shape.kh}x{shape.kw}"
```

Comment:
Wrong indentation here? Format the file? We can also enable pre-commit linting/formatting in this repo (excluding generated files).